learning.py

```diff
@@ -784,15 +785,44 @@ def BackPropagationLearner(dataset, net, learning_rate, epochs, activation=sigmo
                     delta[i] = [leaky_relu_derivative(layer[j].value) * dotproduct(w[j], delta[i+1])
                             for j in range(h_units)]
 
-            #  Update weights
-            for i in range(1, n_layers):
-                layer = net[i]
-                inc = [node.value for node in net[i-1]]
-                units = len(layer)
-                for j in range(units):
-                    layer[j].weights = vector_add(layer[j].weights,
-                                                  scalar_vector_product(
-                                                  learning_rate * delta[i][j], inc))
+            #optimization with velocity gradient
+            t_ = epoch + 1
+            if momentum == True:
+                if epoch == 0:
+                    for i in range(len(delta)):
+                        for j in range(len(delta[i])):
+                            v_dw[i][j] = ((1-beta)*delta[i][j])/(1-beta**(t_+1))
+                else:
+                    for i in range(len(delta)):
+                        for j in range(len(delta[i])):
+                            v_dw[i][j] = (beta*v_dw[i][j]+(1-beta)*delta[i][j])/(1-beta**(t_+1))
+
+
+
+
+
+            #  Update weights without velocity gradient
+            if momentum == False:
+                for i in range(1, n_layers):
+                    layer = net[i]
+                    inc = [node.value for node in net[i-1]]
+                    units = len(layer)
+                    for j in range(units):
+                        layer[j].weights = vector_add(layer[j].weights,
+                                                    scalar_vector_product(
```

Comment:
Again, may be a glitch on how GitHub presents the code, but you should use proper formatting in multi-line code. Like this:

```
function(something,
         something_else(
             other_thing
         )
)
```

Alternatively:

```
function(something,
         something_else(
             other_thing
         ))
```

Reply:
multi-line code style modified